// @vechain/connex.driver-nodejs/dist/driver.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const driver_no_vendor_1 = require("./driver-no-vendor");
const transaction_1 = require("thor-devkit/dist/transaction");
const certificate_1 = require("thor-devkit/dist/certificate");
const blake2b_1 = require("thor-devkit/dist/cry/blake2b");
const crypto_1 = require("crypto");
const options_1 = require("./options");
/** class fully implements Connex.Driver */
class Driver extends driver_no_vendor_1.DriverNoVendor {
    constructor(net, genesis, initialHead, wallet) {
        super(net, genesis, initialHead);
        this.wallet = wallet;
        /** params for tx construction */
        this.txParams = {
            expiration: 18,
            gasPriceCoef: 0
        };
    }
    /**
     * create driver instance
     * it will fetch config(genesis, head) via net as construction params
     * @param net
     * @param wallet
     */
    static connect(net, wallet) {
        return __awaiter(this, void 0, void 0, function* () {
            const genesis = yield net.http('GET', 'blocks/0');
            const best = yield net.http('GET', 'blocks/best', {
                validateResponseHeader: headers => {
                    const xgid = headers['x-genesis-id'];
                    if (xgid && xgid !== genesis.id) {
                        throw new Error(`responded 'x-genesis-id' not matched`);
                    }
                }
            });
            return new Driver(net, genesis, {
                id: best.id,
                number: best.number,
                timestamp: best.timestamp,
                parentID: best.parentID,
                txsFeatures: best.txsFeatures
            }, wallet);
        });
    }
    signTx(msg, option) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.findKey(option.signer);
            const clauses = msg.map(c => ({ to: c.to, value: c.value, data: c.data }));
            const gas = option.gas ||
                (yield this.estimateGas(clauses, key.address));
            const txBody = {
                chainTag: Number.parseInt(this.genesis.id.slice(-2), 16),
                blockRef: this.head.id.slice(0, 18),
                expiration: this.txParams.expiration,
                clauses,
                gasPriceCoef: this.txParams.gasPriceCoef,
                gas,
                dependsOn: option.dependsOn || null,
                nonce: '0x' + crypto_1.randomBytes(8).toString('hex')
            };
            let tx;
            if (option.delegationHandler) {
                const delegatedTx = new transaction_1.Transaction(Object.assign({}, txBody, { reserved: { features: 1 /* vip191 */ } }));
                const originSig = yield key.sign(delegatedTx.signingHash());
                try {
                    const result = yield option.delegationHandler({
                        raw: '0x' + delegatedTx.encode().toString('hex'),
                        origin: key.address
                    });
                    delegatedTx.signature = Buffer.concat([originSig, Buffer.from(result.signature.slice(2), 'hex')]);
                    tx = delegatedTx;
                }
                catch (err) {
                    if (!options_1.options.disableErrorLog) {
                        // tslint:disable-next-line: no-console
                        console.warn('tx delegation error: ', err);
                    }
                    // fallback to non-vip191 tx
                }
            }
            if (!tx) {
                tx = new transaction_1.Transaction(txBody);
                tx.signature = yield key.sign(tx.signingHash());
            }
            const raw = '0x' + tx.encode().toString('hex');
            if (this.onTxCommit) {
                this.onTxCommit({
                    id: tx.id,
                    raw,
                    resend: () => __awaiter(this, void 0, void 0, function* () {
                        yield this.sendTx(raw);
                    })
                });
            }
            yield this.sendTx(raw);
            return {
                txid: tx.id,
                signer: key.address
            };
        });
    }
    signCert(msg, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.findKey(options.signer);
            const annex = {
                domain: 'localhost',
                timestamp: this.head.timestamp,
                signer: key.address
            };
            const unsigned = certificate_1.Certificate.encode(Object.assign({}, msg, annex));
            const signature = yield key.sign(blake2b_1.blake2b256(unsigned));
            return {
                annex,
                signature: '0x' + signature.toString('hex')
            };
        });
    }
    isAddressOwned(addr) {
        return Promise.resolve(this.wallet ? this.wallet.list.findIndex(k => k.address === addr) >= 0 : false);
    }
    findKey(addr) {
        if (this.wallet) {
            const keys = this.wallet.list;
            const key = addr ? keys.find(k => k.address === addr) : keys[0];
            if (key) {
                return key;
            }
        }
        throw new Error('empty wallet');
    }
    sendTx(raw) {
        return this.httpPost('transactions', { raw });
    }
    estimateGas(clauses, caller) {
        return __awaiter(this, void 0, void 0, function* () {
            const outputs = yield this.explain({
                clauses,
                caller,
            }, this.head.id);
            const execGas = outputs.reduce((sum, out) => sum + out.gasUsed, 0);
            const intrinsicGas = transaction_1.Transaction.intrinsicGas(clauses);
            return intrinsicGas + (execGas ? (execGas + 15000) : 0);
        });
    }
}
exports.Driver = Driver;
