// @vechain/connex.driver-nodejs/dist/driver-host.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const WS = require("ws/index");
const json_rpc_1 = require("@vechain/json-rpc");
const options_1 = require("./options");
const methods = [
    'pollHead',
    'getBlock',
    'getTransaction',
    'getReceipt',
    'getAccount',
    'getCode',
    'getStorage',
    'explain',
    'filterEventLogs',
    'filterTransferLogs',
    'signTx',
    'signCert',
    'isAddressOwned',
];
class DriverHost {
    constructor(server, path, acceptor) {
        this.wss = new WS.Server({
            server,
            path
        });
        this.wss.on('connection', (ws, req) => __awaiter(this, void 0, void 0, function* () {
            this.handleConnection(ws, req, acceptor);
        }));
    }
    close() {
        this.wss.close();
    }
    handleConnection(ws, req, acceptor) {
        const rpc = new json_rpc_1.JSONRPC((data, isRequest) => {
            if (!isRequest) {
                data = ' ' + data;
            }
            ws.send(data);
            return Promise.resolve();
        });
        ws.on('message', data => {
            const isRequest = data[0] !== ' ';
            rpc.receive(data, isRequest).catch(err => {
                if (!options_1.options.disableErrorLog) {
                    // tslint:disable-next-line: no-console
                    console.warn('receive jsonrpc payload: ', err);
                }
            });
        });
        let driver;
        rpc.serve(method => {
            if (method === 'connect') {
                return (genesisId) => __awaiter(this, void 0, void 0, function* () {
                    if (driver) {
                        throw new Error('already accepted');
                    }
                    driver = yield acceptor(ws, req, genesisId);
                    return {
                        genesis: driver.genesis,
                        head: driver.head
                    };
                });
            }
            if (methods.includes(method)) {
                return (...args) => {
                    if (!driver) {
                        throw new Error('not accepted');
                    }
                    return driver[method](...args);
                };
            }
        });
    }
}
exports.DriverHost = DriverHost;
