// @vechain/connex.driver-nodejs/dist/simple-wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const secp256k1_1 = require("thor-devkit/dist/cry/secp256k1");
const address_1 = require("thor-devkit/dist/cry/address");
/** class simply implements Wallet interface */
class SimpleWallet {
    constructor() {
        this.keys = [];
    }
    get list() {
        return this.keys.map(k => {
            return {
                address: k.address,
                sign(msgHash) {
                    return Promise.resolve(secp256k1_1.secp256k1.sign(msgHash, k.privateKey));
                }
            };
        });
    }
    /**
     * import private key
     * @param privateKey hex string presented private key
     * @returns address derived from the private key
     */
    import(privateKey) {
        if (privateKey.startsWith('0x')) {
            privateKey = privateKey.slice(2);
        }
        if (!/^[0-9a-f]{64}$/i.test(privateKey)) {
            throw new Error('invalid private key');
        }
        const buf = Buffer.from(privateKey, 'hex');
        const addr = '0x' + address_1.publicKeyToAddress(secp256k1_1.secp256k1.derivePublicKey(buf)).toString('hex');
        this.keys.push({ address: addr, privateKey: buf });
        return addr;
    }
    /**
     * remove corresponding key by given address
     * @param addr address
     * @returns true if found and removed, false otherwise
     */
    remove(addr) {
        const i = this.keys.findIndex(k => k.address === addr.toLowerCase());
        if (i >= 0) {
            this.keys.splice(i, 1);
            return true;
        }
        return false;
    }
}
exports.SimpleWallet = SimpleWallet;
