// @vechain/connex.driver-nodejs/dist/virtual-driver.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const json_rpc_1 = require("@vechain/json-rpc");
const WebSocket = require("isomorphic-ws");
const common_1 = require("./common");
const options_1 = require("./options");
function openWebSocket(url) {
    return new Promise((resolve, reject) => {
        const ws = new WebSocket(url);
        ws.onopen = () => {
            offWebSocket(ws);
            resolve(ws);
        };
        ws.onclose = ev => {
            offWebSocket(ws);
            reject(new Error(`websocket closed: ${ev.reason} (${ev.code})`));
        };
        ws.onerror = ev => {
            offWebSocket(ws);
            reject(ev.error || new Error('websocket error'));
            ws.close();
        };
    });
}
function offWebSocket(ws) {
    ws.onclose = null;
    ws.onerror = null;
    ws.onmessage = null;
    ws.onopen = null;
}
function setupRPC(ws) {
    return __awaiter(this, void 0, void 0, function* () {
        const rpc = new json_rpc_1.JSONRPC((data, isRequest) => {
            if (!isRequest) {
                data = ' ' + data;
            }
            ws.send(data);
            return Promise.resolve();
        });
        ws.onmessage = ev => {
            const isRequest = ev.data[0] !== ' ';
            rpc.receive(ev.data, isRequest)
                .catch(err => {
                if (!options_1.options.disableErrorLog) {
                    // tslint:disable-next-line: no-console
                    console.warn('receive jsonrpc payload: ', err);
                }
            });
        };
        ws.onclose = () => {
            offWebSocket(ws);
            rpc.setError(new Error('closed'));
        };
        ws.onerror = ev => {
            offWebSocket(ws);
            rpc.setError(ev.error);
            ws.close();
        };
        return rpc;
    });
}
function _connect(url, genesisId) {
    return __awaiter(this, void 0, void 0, function* () {
        const ws = yield openWebSocket(url);
        const rpc = yield setupRPC(ws);
        try {
            const info = yield rpc.call('connect', genesisId);
            return {
                ws,
                rpc,
                genesis: info.genesis,
                initHead: info.head
            };
        }
        catch (err) {
            ws.close();
            throw err;
        }
    });
}
function connect(url) {
    return __awaiter(this, void 0, void 0, function* () {
        let conn = yield _connect(url);
        const genesisId = conn.genesis.id;
        const reconnect = () => {
            setTimeout(() => __awaiter(this, void 0, void 0, function* () {
                try {
                    conn = yield _connect(url, genesisId);
                }
                catch (err) {
                    reconnect();
                }
            }), 10 * 1000);
        };
        const rpcCall = (method, ...args) => {
            if (conn) {
                if (conn.ws.readyState === WebSocket.OPEN) {
                    return conn.rpc.call(method, ...args);
                }
                conn = null;
                reconnect();
            }
            return Promise.reject(new Error('closed'));
        };
        let currentHead = conn.initHead;
        return {
            genesis: conn.genesis,
            get head() { return currentHead; },
            pollHead: () => __awaiter(this, void 0, void 0, function* () {
                for (;;) {
                    try {
                        const newHead = yield rpcCall('pollHead');
                        currentHead = newHead;
                        return newHead;
                    }
                    catch (_a) {
                        yield common_1.sleep(5 * 1000);
                    }
                }
            }),
            getBlock: rev => {
                return rpcCall('getBlock', rev);
            },
            getTransaction: id => {
                return rpcCall('getTransaction', id);
            },
            getReceipt: id => {
                return rpcCall('getReceipt', id);
            },
            getAccount: (addr, rev) => {
                return rpcCall('getAccount', addr, rev);
            },
            getCode: (addr, rev) => {
                return rpcCall('getCode', addr, rev);
            },
            getStorage: (addr, key, rev) => {
                return rpcCall('getStorage', addr, key, rev);
            },
            explain: (arg, rev, cacheTies) => {
                return rpcCall('explain', arg, rev, cacheTies);
            },
            filterEventLogs: arg => {
                return rpcCall('filterEventLogs', arg);
            },
            filterTransferLogs: arg => {
                return rpcCall('filterTransferLogs', arg);
            },
            signTx: (msg, opt) => {
                return rpcCall('signTx', msg, opt);
            },
            signCert: (msg, opt) => {
                return rpcCall('signCert', msg, opt);
            },
            isAddressOwned: addr => {
                return rpcCall('isAddressOwned', addr);
            }
        };
    });
}
exports.connect = connect;
