// @vechain/json-rpc/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const payload_1 = require("./payload");
/** Transport-independent JSON RPC 2.0 protocol stack */
class JSONRPC {
    /**
     * create JSON RPC instance
     * @param send the function to transmit JSON RPC payload
     * @param generateId function to generate request id. If omitted, the default  auto-increment function is used.
     */
    constructor(send, generateId) {
        this._ongoings = new Map();
        this._error = null;
        this._send = (payload, isRequest) => {
            if (this._error) {
                return Promise.reject(this._error);
            }
            return send(JSON.stringify(payload), isRequest);
        };
        if (generateId) {
            this._generateId = generateId;
        }
        else {
            let nextId = 0;
            this._generateId = () => {
                return nextId++;
            };
        }
    }
    /**
     * Call to remote method
     * @param method the method name
     * @param params params for method
     * @returns promised method return value
     */
    call(method, ...params) {
        const id = this._generateId();
        const payload = {
            jsonrpc: '2.0',
            id,
            method,
            params
        };
        return new Promise((resolve, reject) => {
            // send request payload, and then add to ongoing list to wait for response
            this._send(payload, true)
                .then(() => {
                this._ongoings.set(id, {
                    resolve,
                    reject
                });
            }).catch(reject);
        });
    }
    /**
     * Send a notification
     * @param method the notification name
     * @param params params for notification
     */
    notify(method, ...params) {
        const payload = {
            jsonrpc: '2.0',
            method,
            params
        };
        return this._send(payload, true);
    }
    /**
     * to receive and process JSON RPC payload
     * @param data JSON encoded payload
     * @param isRequest whether the payload is request type
     */
    receive(data, isRequest) {
        return __awaiter(this, void 0, void 0, function* () {
            let payload;
            try {
                payload = JSON.parse(data);
            }
            catch (err) {
                throw JSONRPC.Error.parseError(err.message);
            }
            try {
                payload_1.Payload.validate(payload, isRequest);
            }
            catch (err) {
                throw JSONRPC.Error.invalidRequest(err.message);
            }
            if (isRequest) {
                yield this._handleRequest(payload);
            }
            else {
                this._handleResponse(payload);
            }
        });
    }
    /**
     * Set to error state, or clear error state.
     * In error state, all subsequent calls will fail immediately.
     * @param err error object
     */
    setError(err) {
        this._error = err;
        if (err) {
            const values = Array.from(this._ongoings.values());
            this._ongoings.clear();
            values.forEach(v => v.reject(err));
        }
    }
    /**
     * serve method handler
     * @param handler handle method calls
     */
    serve(handler) {
        this._handler = handler;
    }
    _handleRequest(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            const impl = this._handler ? this._handler(payload.method) : undefined;
            const hasId = payload.id !== undefined && payload.id !== null;
            if (impl) {
                let result;
                try {
                    result = yield impl(...payload.params);
                }
                catch (err) {
                    if (!hasId) {
                        throw err;
                    }
                    if (err instanceof JSONRPC.Error) {
                        yield this._send(err.asPayload(payload.id), false);
                    }
                    else {
                        yield this._send(JSONRPC.Error.internalError(err.message).asPayload(payload.id), false);
                    }
                    return;
                }
                if (hasId) {
                    yield this._send({
                        jsonrpc: '2.0',
                        id: payload.id,
                        result
                    }, false);
                }
            }
            else {
                const err = JSONRPC.Error.methodNotFound();
                if (hasId) {
                    yield this._send(err.asPayload(payload.id), false);
                }
                else {
                    throw err;
                }
            }
        });
    }
    _handleResponse(payload) {
        const ongoing = this._ongoings.get(payload.id);
        if (!ongoing) {
            return;
        }
        this._ongoings.delete(payload.id);
        if (payload.error) {
            ongoing.reject(new JSONRPC.Error(payload.error.message, payload.error.code, payload.error.data));
        }
        else {
            ongoing.resolve(payload.result);
        }
    }
}
exports.JSONRPC = JSONRPC;
const errorClass = Error;
(function (JSONRPC) {
    class Error extends errorClass {
        constructor(message, code, data) {
            super(message);
            this.code = code;
            this.data = data;
        }
        static parseError(reason) {
            return new Error(`Parse error: ${reason}`, -32700);
        }
        static invalidRequest(reason) {
            return new Error(`Invalid request: ${reason}`, -32600);
        }
        static methodNotFound() {
            return new Error('Method not found', -32601);
        }
        static invalidParams(reason) {
            return new Error(`Invalid params: ${reason}`, -32602);
        }
        static internalError(reason) {
            return new Error(`Internal error: ${reason}`, -32603);
        }
        static serverError(reason, code, data) {
            return new Error(`Server error: ${reason}`, code, data);
        }
        asPayload(id) {
            return {
                jsonrpc: '2.0',
                id,
                error: {
                    code: this.code,
                    message: this.message,
                    data: this.data
                }
            };
        }
    }
    JSONRPC.Error = Error;
})(JSONRPC = exports.JSONRPC || (exports.JSONRPC = {}));
