// @vechain/json-rpc/dist/payload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const V = require("validator-ts");
var Payload;
(function (Payload) {
    const baseScheme = {
        jsonrpc: v => v === '2.0' ? '' : `expected '2.0'`,
        id: V.nullable(V.optional(v => (typeof v === 'number' || typeof v === 'string') ? '' : 'expected number or string')),
        method: () => '',
        params: () => '',
        result: () => '',
        error: () => ''
    };
    const requestScheme = Object.assign(Object.assign({}, baseScheme), { method: V.optional(v => (typeof v === 'string' && v.length > 0) ? '' : 'expected string'), params: v => Array.isArray(v) ? '' : 'expected array' });
    const responseScheme = Object.assign(Object.assign({}, baseScheme), { error: V.optional({
            code: v => typeof v === 'number' ? '' : 'expected number',
            message: v => typeof v === 'string' ? '' : 'expected string',
            data: () => ''
        }) });
    /** validate payload */
    function validate(payload, isRequest) {
        V.validate(payload, isRequest ? requestScheme : responseScheme);
    }
    Payload.validate = validate;
})(Payload = exports.Payload || (exports.Payload = {}));
