// @vechain/connex-framework/dist/event-visitor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const filter_1 = require("./filter");
const rules_1 = require("./rules");
function newEventVisitor(ctx, addr, coder) {
    const encode = (indexed) => {
        const topics = coder.encode(indexed);
        return {
            address: addr,
            topic0: topics[0] || undefined,
            topic1: topics[1] || undefined,
            topic2: topics[2] || undefined,
            topic3: topics[3] || undefined,
            topic4: topics[4] || undefined
        };
    };
    return {
        asCriteria: indexed => {
            try {
                return encode(indexed);
            }
            catch (err) {
                throw new rules_1.BadParameter(`arg0: can not be encoded (${err.message})`);
            }
        },
        filter: (indexed) => {
            rules_1.test(indexed, [{}], 'arg0');
            if (indexed.length === 0) {
                indexed = [{}];
            }
            const criteriaSet = indexed.map((o, i) => {
                try {
                    return encode(o);
                }
                catch (err) {
                    throw new rules_1.BadParameter(`arg0.#${i}: can not be encoded (${err.message})`);
                }
            });
            const filter = filter_1.newFilter(ctx, 'event').criteria(criteriaSet);
            return {
                criteria(set) {
                    filter.criteria(set);
                    return this;
                },
                range(range) {
                    filter.range(range);
                    return this;
                },
                order(order) {
                    filter.order(order);
                    return this;
                },
                apply(offset, limit) {
                    return filter.apply(offset, limit)
                        .then(events => events.map(event => {
                        const decoded = coder.decode(event.data, event.topics);
                        return Object.assign(Object.assign({}, event), { decoded });
                    }));
                }
            };
        }
    };
}
exports.newEventVisitor = newEventVisitor;
