// @vechain/connex-framework/dist/explainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const revert_reason_1 = require("./revert-reason");
const R = require("./rules");
const V = require("validator-ts");
function newExplainer(ctx) {
    const opts = {};
    return {
        caller(addr) {
            opts.caller = R.test(addr, R.address, 'arg0').toLowerCase();
            return this;
        },
        gas(gas) {
            opts.gas = R.test(gas, R.uint64, 'arg0');
            return this;
        },
        gasPrice(gp) {
            opts.gasPrice = R.test(gp, R.bigInt, 'arg0').toString().toLowerCase();
            return this;
        },
        execute(clauses) {
            R.test(clauses, [clauseScheme], 'arg0');
            const transformedClauses = clauses.map(c => {
                return {
                    to: c.to ? c.to.toLowerCase() : null,
                    value: c.value.toString().toLowerCase(),
                    data: (c.data || '0x').toLowerCase()
                };
            });
            return ctx.driver.explain(Object.assign({ clauses: transformedClauses }, opts), ctx.trackedHead.id)
                .then(outputs => {
                return outputs.map(o => {
                    if (o.reverted) {
                        const revertReason = revert_reason_1.decodeRevertReason(o.data);
                        return Object.assign(Object.assign({}, o), { decoded: { revertReason } });
                    }
                    return o;
                });
            });
        }
    };
}
exports.newExplainer = newExplainer;
const clauseScheme = {
    to: V.nullable(R.address),
    value: R.bigInt,
    data: V.optional(R.bytes)
};
