// @vechain/connex-framework/dist/method.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const revert_reason_1 = require("./revert-reason");
const R = require("./rules");
function newMethod(ctx, addr, coder) {
    let value = 0;
    const opts = {};
    let cacheTies;
    return {
        value(val) {
            value = R.test(val, R.bigInt, 'arg0');
            return this;
        },
        caller(caller) {
            opts.caller = R.test(caller, R.address, 'arg0').toLowerCase();
            return this;
        },
        gas(gas) {
            opts.gas = R.test(gas, R.uint64, 'arg0');
            return this;
        },
        gasPrice(gp) {
            opts.gasPrice = R.test(gp, R.bigInt, 'arg0').toString().toLowerCase();
            return this;
        },
        cache(ties) {
            cacheTies = R.test(ties, [R.address], 'arg0').map(t => t.toLowerCase());
            return this;
        },
        asClause: (...args) => {
            const inputsLen = (coder.definition.inputs || []).length;
            R.ensure(inputsLen === args.length, `args count expected ${inputsLen}`);
            try {
                const data = coder.encode(...args);
                return {
                    to: addr,
                    value: value.toString().toLowerCase(),
                    data
                };
            }
            catch (err) {
                throw new R.BadParameter(`args can not be encoded (${err.message})`);
            }
        },
        call(...args) {
            const clause = this.asClause(...args);
            return ctx.driver.explain(Object.assign({ clauses: [clause] }, opts), ctx.trackedHead.id, cacheTies)
                .then(outputs => outputs[0])
                .then(output => {
                if (output.reverted) {
                    const revertReason = revert_reason_1.decodeRevertReason(output.data);
                    return Object.assign(Object.assign({}, output), { decoded: { revertReason } });
                }
                else {
                    const decoded = coder.decode(output.data);
                    return Object.assign(Object.assign({}, output), { decoded });
                }
            });
        }
    };
}
exports.newMethod = newMethod;
