// @vechain/connex-framework/dist/thor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const account_visitor_1 = require("./account-visitor");
const block_visitor_1 = require("./block-visitor");
const tx_visitor_1 = require("./tx-visitor");
const filter_1 = require("./filter");
const head_tracker_1 = require("./head-tracker");
const explainer_1 = require("./explainer");
const R = require("./rules");
function newThor(driver) {
    const headTracker = head_tracker_1.newHeadTracker(driver);
    const ctx = {
        driver,
        get trackedHead() { return headTracker.head; }
    };
    const genesis = JSON.parse(JSON.stringify(driver.genesis));
    return {
        get genesis() { return genesis; },
        get status() {
            return {
                head: headTracker.head,
                progress: headTracker.progress
            };
        },
        ticker: () => headTracker.ticker(),
        account: addr => {
            addr = R.test(addr, R.address, 'arg0').toLowerCase();
            return account_visitor_1.newAccountVisitor(ctx, addr);
        },
        block: revision => {
            if (typeof revision === 'undefined') {
                revision = ctx.trackedHead.id;
            }
            else {
                R.ensure(typeof revision === 'string' ? R.isHexBytes(revision, 32) : R.isUInt(revision, 32), 'arg0: expected bytes32 or unsigned 32-bit integer');
            }
            return block_visitor_1.newBlockVisitor(ctx, typeof revision === 'string' ? revision.toLowerCase() : revision);
        },
        transaction: id => {
            id = R.test(id, R.bytes32, 'arg0').toLowerCase();
            return tx_visitor_1.newTxVisitor(ctx, id);
        },
        filter: kind => {
            R.ensure(kind === 'event' || kind === 'transfer', `arg0: expected 'event' or 'transfer'`);
            return filter_1.newFilter(ctx, kind);
        },
        explain: () => explainer_1.newExplainer(ctx)
    };
}
exports.newThor = newThor;
