// @vechain/connex-framework/dist/vendor.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const R = require("./rules");
const V = require("validator-ts");
const abi_1 = require("thor-devkit/dist/abi");
function newVendor(driver) {
    return {
        sign: (kind) => {
            if (kind === 'tx') {
                return newTxSigningService(driver);
            }
            else if (kind === 'cert') {
                return newCertSigningService(driver);
            }
            else {
                throw new R.BadParameter(`arg0: expected 'tx' or 'cert'`);
            }
        },
        owned: (addr) => {
            addr = R.test(addr, R.address, 'arg0').toLowerCase();
            return driver.isAddressOwned(addr);
        }
    };
}
exports.newVendor = newVendor;
function newTxSigningService(driver) {
    const opts = {};
    return {
        signer(addr) {
            opts.signer = R.test(addr, R.address, 'arg0').toLowerCase();
            return this;
        },
        gas(gas) {
            opts.gas = R.test(gas, R.uint64, 'arg0');
            return this;
        },
        dependsOn(txid) {
            opts.dependsOn = R.test(txid, R.bytes32, 'arg0').toLowerCase();
            return this;
        },
        link(url) {
            opts.link = R.test(url, R.string, 'arg0');
            return this;
        },
        comment(text) {
            opts.comment = R.test(text, R.string, 'arg0');
            return this;
        },
        delegate(handler) {
            R.ensure(typeof handler === 'function', `arg0: expected function`);
            opts.delegationHandler = (unsigned) => __awaiter(this, void 0, void 0, function* () {
                const obj = yield handler(unsigned);
                R.test(obj, {
                    signature: v => R.isHexBytes(v, 65) ? '' : 'expected 65 bytes'
                }, 'delegation-result');
                return {
                    signature: obj.signature.toLowerCase()
                };
            });
            return this;
        },
        request(msg) {
            R.test(msg, [clauseScheme], 'arg0');
            const transformedMsg = msg.map(c => ({
                to: c.to ? c.to.toLowerCase() : null,
                value: c.value.toString().toLowerCase(),
                data: (c.data || '0x').toLowerCase(),
                comment: c.comment,
                abi: c.abi ? JSON.parse(JSON.stringify(c.abi)) : c.abi
            }));
            return (() => __awaiter(this, void 0, void 0, function* () {
                try {
                    return yield driver.signTx(transformedMsg, opts);
                }
                catch (err) {
                    throw new Rejected(err.message);
                }
            }))();
        }
    };
}
function newCertSigningService(driver) {
    const opts = {};
    return {
        signer(addr) {
            opts.signer = R.test(addr, R.address, 'arg0').toLowerCase();
            return this;
        },
        link(url) {
            opts.link = R.test(url, R.string, 'arg0');
            return this;
        },
        request(msg) {
            R.test(msg, {
                purpose: v => (v === 'agreement' || v === 'identification') ?
                    '' : `expected 'agreement' or 'identification'`,
                payload: {
                    type: v => v === 'text' ? '' : `expected 'text'`,
                    content: R.string
                }
            }, 'arg0');
            return (() => __awaiter(this, void 0, void 0, function* () {
                try {
                    return yield driver.signCert(msg, opts);
                }
                catch (err) {
                    throw new Rejected(err.message);
                }
            }))();
        }
    };
}
class Rejected extends Error {
    constructor(msg) {
        super(msg);
    }
}
Rejected.prototype.name = 'Rejected';
const clauseScheme = {
    to: V.nullable(R.address),
    value: R.bigInt,
    data: V.optional(R.bytes),
    comment: V.optional(R.string),
    abi: V.optional(v => {
        if (!(v instanceof Object)) {
            return 'expected object';
        }
        try {
            // tslint:disable-next-line: no-unused-expression
            new abi_1.abi.Function(v);
            return '';
        }
        catch (err) {
            return `expected valid ABI (${err.message})`;
        }
    })
};
